Front-end, subxt and libraries (#39)
substrate-developer-hub/docs-sandbox: source/docs/reference/command-line-tools/subxt-cli.md

# subxt-cli

A CLI utility to generate the metadata and API from a running Substrate node.
This tool was originally designed for the [`subxt` library](./06-build/libraries#subxt) to build the runtime API from the metadata of a target node. 

> It can be used as a standalone tool to download the metadata or generate the API for the runtime of any node. 

Reviewer: Currently it only supports downloading the metadata for nodes running V14 or later, since it is geared towards acquiring the metadata for the `subxt` library to consume (which only supports V14+)

## Installation 

To install `subxt-cli`:

1. Open a terminal shell, if necessary.
1. Verify that you have the Rust compiler and toolchain, if necessary.
1. Download the required packages with the following command: 

    `cargo install subxt-cli`

## Basic command usage

The basic syntax for running `subxt-cli` commands is:

`subxt <SUBCOMMAND>`

> In order to use the commands exposed by `subxt`, you must either be running a node locally or you must specify the chain you're targeting.

Reviewer: Specifically the `metadata` command, anyway. the `codegen` command can talk to a node, or use the metadata already acquired via the `metadata` command.

### Flags

You can use the following optional flags with the `subxt` command.

| Flag | Description
| ------- | -----------
| -h, --help | Displays usage information. 
| -V, --version | Displays version information.

### Subcommands

You can use the following subcommands with the `subxt` command. 
For reference information and examples that illustrate using `subxt` subcommands, select an appropriate command.

| Command | Description
| ------- | -----------
| [`codegen`](#codegen) | Generate runtime API client code from metadata.  
| [`metadata`](#metadata) | Download metadata from a Substrate node, for use with `subxt` codegen.

### Output

Depending on how you specify the subcommand, the output from `subxt` displays some or all of the following information:

| This field | Contains
| ---------- | ----------
| Metadata | A file with the metadata of a target chain.
| API | A file with the API of the target chain.

### Examples

To display version information for the `subxt` program, run the following command:

`subxt --version`

To display usage information for the `subxt metadata` command, run the following command:

`subxt metadata --help`

## codegen

> Use the `subxt codegen` command to generate a the API for some target Substrate node.

Reviewer: ```suggestion
Use the `subxt codegen` command to generate an interface for some target Substrate node.
```

Might be worth mentioning that this is analogous to what the `subxt` macro does.

This could be useful for debugging or modifying a node's API to meet certain hardware constraints.

#### Basic usage

`subxt codegen [OPTIONS]`

#### Flags

You can use the following optional flags with the `subxt codegen` command.

| Flag   | Description
| ------ | -----------
| `-h, --help`  | Displays usage information.
| `-V, --version` | Prints version information.

#### Options

You can use the following command-line options with the `subxt codegen` command.

| Option   | Description
| -------- | -----------
| `-f, --file <file>` | The path to the encoded metadata file.
| `--url <url>` | The url of the Substrate node to query for metadata for codegen.

#### Examples

To format the generated API and print it to the terminal:

> `subxt codegen | rustfmt --edition=2018 --emit=stdout`

Reviewer: Just `rustfmt` on its own works well for me (and I don't see why the `--edition` flag is necessary, but if you'd like to keep it, perhaps point at the `2021` edition?)

Author: I actually get this error if I don't specify an edition that's >2018:

```bash
error[E0670]: `async fn` is not permitted in Rust 2015
```

That not happen with you ?
